flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/interval/asyncprocess/AsyncTimeIntervalJoin.java

```diff
@@ -0,0 +1,345 @@
+
+        paddingUtil = new OuterJoinPaddingUtil(leftType.toRowSize(), rightType.toRowSize());
+    }
+
+    @Override
+    public void onTimer(long timestamp, OnTimerContext ctx, Collector<RowData> out)
+            throws Exception {
+        joinFunction.setJoinKey(ctx.getCurrentKey());
+        joinCollector.setInnerCollector(out);
+        updateOperatorTime(ctx);
+        // In the future, we should separate the left and right watermarks. Otherwise, the
+        // registered timer of the faster stream will be delayed, even if the watermarks have
+        // already been emitted by the source.
+        leftTimerState
+                .asyncValue()
+                .thenAccept(
+                        leftCleanUpTime -> {
+                            if (leftCleanUpTime != null && timestamp == leftCleanUpTime) {
+                                rightExpirationTime =
+                                        calExpirationTime(leftOperatorTime, rightRelativeSize);
+                                removeExpiredRows(
+                                        rightExpirationTime,
+                                        rightCache,
+                                        leftTimerState,
+                                        ctx,
+                                        false);
+                            }
+                        });
+
+        rightTimerState
+                .asyncValue()
+                .thenAccept(
+                        rightCleanUpTime -> {
+                            if (rightCleanUpTime != null && timestamp == rightCleanUpTime) {
+                                leftExpirationTime =
+                                        calExpirationTime(rightOperatorTime, leftRelativeSize);
+                                removeExpiredRows(
+                                        leftExpirationTime, leftCache, rightTimerState, ctx, true);
+                            }
+                        });
+    }
+
+    @Override
+    protected StateFuture<Boolean> iterateRightCache(
+            RowData leftRow, long rightQualifiedLowerBound, long rightQualifiedUpperBound)
+            throws Exception {
+        Set<Long> toRemove = new HashSet<>();
+        final AtomicBoolean emitted = new AtomicBoolean(false);
+        CompletableFuture<Boolean> emitResult = new CompletableFuture<>();
+
+        Iterator<Map.Entry<Long, List<Tuple2<RowData, Boolean>>>> rightIterator =
+                rightCache.iterator();
+        return rightCache
+                .asyncEntries()
+                .thenCompose(
+                        iter ->
+                                iter.onNext(
+                                        entry -> {
+                                            boolean entryEmitted =
+                                                    tryMatchLeftRowsByTime(
+                                                            leftRow,
+                                                            rightIterator.next(),
```

Comment:
```suggestion
                                                            entry,
```

And remove the `rightIterator` above